### PR TITLE
Noise generators: Perlin and interpolated

### PR DESCRIPTION
**Pull request by xymus**
Intro the PerlinNoise generator and the underlying (but independent) InterpolatedNoise, with some related services.

The rosetta code Perlin noise example (in 3D) is for comparison to my implementation in `lib/noise.nit` in (2D). We do not get the expected result with this implementation, 0.13691995878400010 vs 0.13691995878400012. It may be an error in the implementation of `Float.to_precision`... this will require further investigation.

The optimized `SequenceRead::rand` is a bonus! it is not used in this PR.

### CODE REVIEW

File: lib/noise.nit [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Provides the noise generators `PerlinNoise` and `InterpolatedNoise`
+module noise
+
+# 2D noise generator
+abstract class Noise
+
+	# Lowest possible value returned by `[]`
```
Comment (privat) [inline lib/noise.nit]: C'est un peu bizarre de référencer [] alors que c'est la dernière fonction. Et le litterate programming alors?

File: lib/noise.nit [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	#
+	# Default at `1.0`.
```
Comment (privat) [inline lib/noise.nit]: Mensonges !

File: lib/noise.nit [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+#     end
+#     print ""
+# end
+# assert max <= map.max
+# assert min >= map.min
+# ~~~
+#
+# ## Result at seed == 0
+#
+# ~~~
```
Comment (privat) [inline lib/noise.nit]: `~~~raw`

File: lib/noise.nit [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+class PerlinNoise
+	super Noise
+
+	# Desired number of `layers`
+	#
+	# This attribute must be assigned before any call to `layers` or `[]`.
+	#
+	# By default, it is the highest integer under the logarithm base 2
+	# of `amplitude`.
+	var n_layers: Int = amplitude.log_base(2.0).to_i is lazy, writable
```
Comment (privat) [inline lib/noise.nit]: `to_i` truncates or rounds ? Just to be sure about the comment that states `under`

File: examples/rosettacode/perlin_noise.nit [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+# This program is public domain
+
+# Task: Perlin noise
+#
+# See: <http://rosettacode.org/wiki/Perlin_noise>
+module perlin_noise
+
+redef universal Float
+	# Smoothened `self`
+	fun fade: Float do return self*self*self*(self*(self*6.0-15.0)+10.0)
```
Comment (privat) [inline examples/rosettacode/perlin_noise.nit]: gné?

**Comment by privat** [comment]
+1

File: examples/rosettacode/perlin_noise.nit (hunk 5 above)
Comment (xymus) [inline examples/rosettacode/perlin_noise.nit]: Don't ask me... I followed the reference implementation as closely as possible.

**Comment by xymus** [comment]
Rerolled with @privat's comments and fixing tests. Improved the doc of the classes in the noise module. And revisited default value of `PerlinNoise::n_levels` so it has a minimum of 4.

File: examples/rosettacode/perlin_noise.nit (hunk 5 above)
Comment (Morriar) [inline examples/rosettacode/perlin_noise.nit]: lol...

File: lib/noise.nit [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Provides the noise generators `PerlinNoise` and `InterpolatedNoise`
+module noise
+
+# 2D noise generator
+abstract class Noise
```
Comment (Morriar) [inline lib/noise.nit]: I find this funny that the class that generates noise is called noise. Why not NoiseGenerator or NoiseGenerator2D?

File: lib/noise.nit [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+# 7888999988888888888777666666666777777777776666666666666667777788899988
+# 7888999988888887777776666666666777777777776666766666666677788888899988
+# 7788999998888877777776666666666777777777776667776666667778888888899988
+# 7788999988888777777776666666666777776677766777776677777788888889999988
+# 6788999988888777777766666666666677676677667777766677777888999999999998
+# 7788888888888777777776666666666666676677777776666777778889999999999998
+# 6778888888888777777766666566666666667777666776667777888999999999999999
+# 6778888888888777777766666566666666667776666776667778889999999999999999
+# 6778888888777777777666666566666666667776666777677788899999999999888999
+# 6777787777777777776666666666666666677777677777777888999999999998888899
```
Comment (Morriar) [inline lib/noise.nit]: I was expecting for fancy ASCII art as an easter egg. I'm disapointed...

**Comment by Morriar** [comment]
+1

File: lib/noise.nit (hunk 7 above)
Comment (xymus) [inline lib/noise.nit]: I was looking for pretty ones at first... but I settled on those at seed = 0 for simplicity. :disappointed:

File: lib/noise.nit (hunk 6 above)
Comment (xymus) [inline lib/noise.nit]: At first it was `NoiseMap` but it quickly became redundant. Especially with `InterpolatedNoiseMap` where the `Map` part didn't convey much information. We'll have to revisit these names if/when we add the 1D, 3D and 4D versions.

**Comment by R4PaSs** [comment]
+1

**Pull request merged**